Design/ISIS_Imaging/ISIS_Imaging_general.md

```diff
@@ -0,0 +1,182 @@
+
+To answer the initial question, building the selection for the user is now just a matter of iterating the dict and getting the (user friendly) name, not the uuid, and displaying that.
+
+The unique ID will be generated using python's `uuid` package.
+
+A future upgrade on this will take into account which is the _current focus_ of the user, and when applying an operation will automatically select _that_ stack as the default choice.
+
+## Visualisation
+Visualisation will be done using Matplotlib. Each stack will be a separate part of a QDockWidget, so the users will be able to arrange the visualised stacks in any way they want. This also means that there can be more than one stack loaded at a time.
+
+This must allow for the user to go to any indice from the volume. 
+
+The visualisation must allow for a rectangle ROI selection, that is persistent if the image underneath is changed. It also must allow for the visualisation of a histogram of the selected ROI (computation should be done in the `core` package).
+
+Currently the idea is to use `FigureCanvasQtAgg` with the default drawing algorithms, for visusalisation of the images. Matplotlib provides a `RectangleSelector` class which will be used to do the ROI selection. If this proves to be too slow, a solution may be to tailor the drawing function, while still retaining the ROI selection functionality.
+
+## Applying a filter
+Applying a filter will bring up a dialogue in which the user has to select on which stack to apply the filter via a dropdown menu, and fill in the required parameters the filter has.
+
+Any filters should be dynamically registered using a [cli_registrator](https://github.com/mantidproject/isis_imaging/blob/master/isis_imaging/core/algorithms/cli_registrator.py) style approach.
+
+Current issues for this section:
+- Dynamic dialogue building
+  - Solution is to use an approach like the `cli_registrator`, but pass in the QObject in which the filters will register themselves in
+  - Dialogue registration is a bit more complicated, because we have to connect the dropdown menu to the dialogue's `.show()` method.
+- Transferring information from the dialogue (the parameters) to the execution
+  - Current solution is to use a partial function, decorated with all of the parameters. No other function down the chain of execution will have to worry about the parameters that way.
+- Handling filters with multiple steps
+  - Background correction and contrast normalisation need to calculate values for scaling, do the filter execution, and then apply the scaling to the images.
+     - Currently the solution is to apply a monad style approach.
+
+## Undoing an operation
+Building the Undo operation may be complicated, but here are a few high level approaches:
```

Comment:
Could you store each image operation as a list and only apply to the image at the point of visualization? Then you could pop operations off the list (instead of managing crazy amounts of memory). Apologies if this is basically what you were trying to say here.